core/sequencer/server.go

```diff
@@ -100,13 +100,41 @@ type Batcher interface {
 	ReadBatch(ctx context.Context, directoryID string, rev int64) (*spb.MapMetadata, error)
 }
 
+// MapWriter writes to the trillian Map.
+type MapWriter interface {
+	// WriteMap sends leaves to the map for directoryID.
+	WriteMap(ctx context.Context, leaves []*tpb.MapLeaf, meta *spb.MapMetadata, directoryID string) error
+}
+
+// MapReader reads a set of indexes from the map.
+type MapReader interface {
+	// ReadMap emits one map leaf for every index requested.
+	ReadMap(ctx context.Context, indexes [][]byte, directoryID string, rev int64,
```

Comment:
Prereq: see "DoFn and other user functions" -> "(2) Struct with methods" and DoFn [godoc](https://godoc.org/github.com/apache/beam/sdks/go/pkg/beam#hdr-DoFns). Apparently, Beam API allows `DoFn` be a struct implementing their interface.

An alternative approach to yours is to rename `ReadMap` to `ProcessElement`, and drop `directoryID` and `rev` from this signature. Sending these 2 parameters via side input doesn't look reasonable to me as those are construction time parameters for the pipeline. Instead you can assume that `MapReader` implementations know about `directoryID` and `rev`.

If you do that then `MapReaderFn` could be a better name.

Reply:
Same reasoning applies to all interfaces in this file.